app/coffee/RoomManager.coffee

```diff
@@ -88,41 +90,36 @@ module.exports = RoomManager =
+        RoomClients.set(room, []) unless RoomClients.has(room)
+        RoomClients.get(room).push(client)
+
+    _clientLeaveRoom: (client, room) ->
+        client.rooms.splice(client.rooms.indexOf(room), 1)
+        clients = RoomClients.get(room)
+        clients.splice(clients.indexOf(client), 1)
+        if clients.length == 0
+            RoomClients.delete(room)
 
     _roomsClientIsIn: (client) ->
-        return @getRoomsClientIsInSync(client)
+        return client.rooms.slice()
 
     _clientAlreadyInRoom: (client, room) ->
-        return @getRoomsClientIsInSync(client).indexOf(room) != -1
-
-    getClientsInRoomSync: (io, room) ->
-        # the implementation in socket.io-adapter is prone to race conditions:
-        # it passes the list of clients via process.nextTick, but given that
-        # NodeJS can process multiple network events in the same event loop
-        # cycle, multiple clients may seem to be the last in a room, but
-        # actually there are not -- some other client joined in the mean time.
-        # (mean time: calculating the list of clients and us receiving it)
-        adapter = io.sockets.adapter
-        return [] unless adapter.rooms.hasOwnProperty(room)
-        return Object.keys(adapter.rooms[room].sockets).filter((id) ->
-          return adapter.nsp.connected[id]
-        )
+        return client.rooms.includes(room)
+
+    # UNSAFE: iterating over the client list returned from here is not safe
+    # e.g. calling .disconnect() on a list member will manipulate the list and
+    #       hence break the iterator.
+    _getClientsInRoomSyncUnsafe: (room) ->
+        return RoomClients.get(room) || []
+
+    countClientsInRoom: (room) ->
+        return RoomManager._getClientsInRoomSyncUnsafe(room).length
+        
+    getClientsInRoomSync: (room) ->
+        return RoomManager._getClientsInRoomSyncUnsafe(room).slice()
 
     # HACK: it calls the callback synchronously -- hence the name pseudoAsync
-    # calling it asynchronously would lead to race conditions -- see above
-    getClientsInRoomPseudoAsync: (io, room, cb) ->
-        cb(null, RoomManager.getClientsInRoomSync(io, room))
-
-    getRoomsClientIsInSync: (client) ->
-        # The implementation in socket.io is prone to race conditions:
-        # The adapter will get updated immediately from joining, but the
-        #  client local state (.rooms) is updated via process.nextTick.
-        # Given that NodeJS can process multiple network events in the same
-        #  event loop cycle, we can join a project and disconnect in the same
-        #  cycle. We would not pick up the room (we just joined) for leaving.
-        roomRegistry = client.server.sockets.adapter.sids
-        return [] unless roomRegistry.hasOwnProperty(client.id)
-        rooms = (room for room in Object.keys(roomRegistry[client.id]) when room isnt client.id)  # exclude the client socket entry
-        return rooms
+    getClientsInRoomPseudoAsync: (room, cb) ->
```

Comment:
`getClientsInRoomSnapshot` might be a better name for this?

Reply:
Note that this name is not new, only the signature changed (like many others -- I should have kept the `io` parameter for a simpler review). I agree that the name is not perfect. A refactoring in the calling logic to a synchronous call would eliminate it.